File: 23mergeKSortedLists/mergeKSortedLists.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    public ListNode mergeKLists(ListNode[] lists) {
        final var result = new ListNode(0);
        var current = result;
        for (final var $: (Iterable<Integer>)Stream.of(lists).flatMap(list -> Stream.iterate(list, Objects::nonNull, $ -> $.next)).mapToInt($ -> $.val).sorted()::iterator)
        {
            current.next = new ListNode($);
            current = current.next;
        }
        return result.next;
    }
}

File: 23mergeKSortedLists/mergeKSortedLists.js
/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */
/**
 * @param {ListNode[]} lists
 * @return {ListNode}
 */

const lodash = require('lodash')

ListNode.prototype[Symbol.iterator] = function*()
{
    let _ = this
    while (_)
    {
        yield _
        _ = _.next
    }
}

var mergeKLists = function(lists) {
    const result = new ListNode(null)
    let current = result
    for (const _ of lodash.chain(lists).filter(_ => !Object.is(_, null)).flatMap(lodash.unary(Array.from)).sortBy([_ => _.val]))
    {
        current.next = _
        current = current.next
    }
    return result.next
};
